File: src/SWA.Ariadne.Gui/Mazes/ImageLoader.cs
using System;
using System.Collections.Generic;
using System.Collections;
using System.Text;
using System.Drawing;
using System.Threading;
using SWA.Utilities;

namespace SWA.Ariadne.Gui.Mazes
{
    public class ImageLoader
    {
        #region Member variables.

        /// <summary>
        /// This directory and its subdirectories are searched for images to be loaded.
        /// </summary>
        private readonly string imageFolder;

        /// <summary>
        /// Minimum and maximum size (in each dimension) the images should be scaled to.
        /// </summary>
        private readonly int minSize, maxSize;

        /// <summary>
        /// Synchronized queue for placing the loaded images.
        /// </summary>
        private readonly Queue queue;

        /// <summary>
        /// Desired number of images that should be in the queue.
        /// </summary>
        private readonly int queueLength;

        /// <summary>
        /// Background thread responsible for filling the queue.
        /// </summary>
        private readonly Thread thread;

        /// <summary>
        /// Used for regulating the buffered queue length.
        /// </summary>
        private readonly Semaphore queueEmptySemaphore, queueFullSemaphore;

        /// <summary>
        /// A list of recently used images.  We'll try to avoid using the same images in rapid succession.
        /// </summary>
        private List<string> recentlyUsedImages = new List<string>();

        #endregion

        #region Constructor.

        /// <summary>
        /// Constructor.
        /// </summary>
        /// <param name="minSize"></param>
        /// <param name="maxSize"></param>
        /// <param name="imageFolder"></param>
        /// <param name="queueLength">Number of images to be pre-loaded into a queue.  If positive, the loader uses a background thread.</param>
        public ImageLoader(int minSize, int maxSize, string imageFolder, int queueLength)
        {
            this.minSize = minSize;
            this.maxSize = maxSize;
            this.imageFolder = imageFolder;
            this.queueLength = queueLength;

            if (queueLength > 0)
            {
                this.queue = Queue.Synchronized(new Queue(queueLength));
                this.thread = new Thread(new ThreadStart(LoadImages));
                this.queueEmptySemaphore = new Semaphore(0, queueLength);
                this.queueFullSemaphore = new Semaphore(queueLength, queueLength);

                thread.Priority = ThreadPriority.BelowNormal;
                thread.Start();
            }
        }

        /// <summary>
        /// Call this method when the application is going to terminate.
        /// </summary>
        public void Shutdown()
        {
            if (this.thread != null)
            {
                thread.Abort();
                thread.Join();
            }
        }

        #endregion

        #region Access methods.

        /// <summary>
        /// Returns the next queued image.
        /// Returns null if no valid images can be found.
        /// </summary>
        /// <param name="r"></param>
        /// <returns></returns>
        public ContourImage GetNext(Random r)
        {
            ContourImage result;

            if (queueLength > 0)
            {
                queueEmptySemaphore.WaitOne();
                result = queue.Dequeue() as ContourImage;
                queueFullSemaphore.Release();
            }
            else
            {
                List<string> imagePaths = FindImages(imageFolder, 1, false, r);
                if (imagePaths.Count == 0)
                {
                    return null;
                }
                string imagePath = imagePaths[0];
                result = LoadImage(imagePath, r);
                result.ProcessImage();
            }

            return result;
        }

        #endregion

        #region Image provider thread.

        /// <summary>
        /// Background thread:
        /// Fills the queue with images.
        /// </summary>
        private void LoadImages()
        {
            Random r = RandomFactory.CreateRandom();

            #region Start with a few images, preferrably without a contour.

            List<ContourImage> unprocessedImages = new List<ContourImage>(queueLength + 1);

            foreach (string imagePath in FindImages(imageFolder, queueLength + 1, true, r))
            {
                ContourImage img = LoadImage(imagePath, r);
                if (img == null)
                {
                    // continue;
                }
                else if (img.HasContour)
                {
                    unprocessedImages.Add(img);
                }
                else
                {
                    queueFullSemaphore.WaitOne();
                    queue.Enqueue(img);
                    queueEmptySemaphore.Release();
                }
            }

            foreach (ContourImage img in unprocessedImages)
            {
                queueFullSemaphore.WaitOne();
                img.ProcessImage();
                queue.Enqueue(img);
                queueEmptySemaphore.Release();
            }

            #endregion

            #region Continuously load more images, keeping the queue full.

            while (true)
            {
                int loadedImagesCount = 0;

                foreach (string imagePath in FindImages(imageFolder, 100, false, r))
                {
                    ContourImage img = LoadImage(imagePath, r);
                    if (img == null)
                    {
                        continue;
                    }

                    // Wait while the queue contains enough items.
                    queueFullSemaphore.WaitOne();

                    // Enqueue the processed image.
                    img.ProcessImage();
                    queue.Enqueue(img);
                    queueEmptySemaphore.Release();
                    
                    ++loadedImagesCount;
                }

                // If no image was loaded successfully, enqueue a null value.
                if (loadedImagesCount == 0)
                {
                    queueFullSemaphore.WaitOne();
                    queue.Enqueue(null);
                    queueEmptySemaphore.Release();
                }
            }

            #endregion
        }

        /// <summary>
        /// Returns an image loaded from the given path and scaled to the known size range.
        /// </summary>
        /// <param name="imagePath"></param>
        /// <param name="r"></param>
        /// <returns></returns>
        private ContourImage LoadImage(string imagePath, Random r)
        {
            try
            {
                Image img = new Bitmap(imagePath);

                #region Scale img so that its larger dimension is between the desired min and max size.

                if (img.Width > maxSize || img.Height > maxSize)
                {
                    int d = r.Next(minSize, maxSize);
                    int h = img.Height, w = img.Width;
                    if (h > w)
                    {
                        w = d * w / h;
                        h = d;
                    }
                    else
                    {
                        h = d * h / w;
                        w = d;
                    }
                    img = new Bitmap(img, new Size(w, h));
                }

                #endregion

                return new ContourImage(img);
            }
            catch (Exception e)
            {
                System.Console.Out.WriteLine("failed loading image [{0}]: {1}", imagePath, e.ToString());
                return null;
            }
        }

        /// <summary>
        /// Returns a list of file names in or below the given path.
        /// Only the following image file types are considered: JPG, GIF, PNG.
        /// </summary>
        /// <param name="folderPath"></param>
        /// <param name="count"></param>
        /// <param name="quickSearch">when true, only JPG files are considered (if there are at least 100 of them)</param>
        /// <returns></returns>
        private List<string> FindImages(string folderPath, int count, bool quickSearch, Random r)
        {
            if (folderPath == null || count < 1)
            {
                return new List<string>();
            }

            List<string> availableImages = new List<string>();

            availableImages.AddRange(SWA.Utilities.Directory.Find(folderPath, "*.jpg", true));

            if (quickSearch == false || availableImages.Count < 100)
            {
                availableImages.AddRange(SWA.Utilities.Directory.Find(folderPath, "*.gif", true));
                availableImages.AddRange(SWA.Utilities.Directory.Find(folderPath, "*.png", true));
            }

            List<string> result = new List<string>(count);

            // Shorten the list of recently used images.
            // Make sure the list does not get too short.
            while (this.recentlyUsedImages.Count > 0
                && this.recentlyUsedImages.Count > availableImages.Count - count
                && this.recentlyUsedImages.Count > availableImages.Count * 3 / 4
                )
            {
                // Remove an item near the beginning of the list (recently added items are at the end).
                this.recentlyUsedImages.RemoveAt(r.Next(this.recentlyUsedImages.Count / 3 + 1));
            }

            // Select the required number of images.
            // Avoid recently used images.
            while (result.Count < count && availableImages.Count > 0)
            {
                int p = r.Next(availableImages.Count);
                string imagePath = availableImages[p];

                if (!recentlyUsedImages.Contains(imagePath))
                {
                    result.Add(imagePath);
                    recentlyUsedImages.Add(imagePath);
                }

                availableImages.RemoveAt(p);
            }

            return result;
        }

        #endregion
    }
}

File: src.JAVA/src/swa/ariadne/gui/images/ImageLoader.java
package swa.ariadne.gui.images;

import java.awt.image.BufferedImage;
import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;
import java.util.Queue;
import java.util.Random;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.Semaphore;

import javax.imageio.ImageIO;

import swa.ariadne.settings.RegisteredOptions;
import swa.util.IParameterFile;
import swa.util.ImageUtil;
import swa.util.IntRange;
import swa.util.RandomFactory;

/**
 * The ImageLoader can be used to pre-fetch several images in a background thread
 * and convert them to {@link ContourImage ContourImages}.
 * <br>
 * The {@link #getNext(Random)} method will immediately return the next available image,
 * or wait until the next conversion is finished.
 *
 * @author Stephan.Wacker@web.de
 */
public final
class ImageLoader
implements IImageLoader
{
    //--------------------- Constants

    /** Used to separate several image paths. */
    private final String PathSeparator = ";";

    //--------------------- Member variables and Properties

    /** This directory and its sub-directories are searched for images to be loaded. */
    private final String imageFolder;

    /** Minimum and maximum size (in each dimension) the images should be scaled to. */
    private final IntRange sizeRange = new IntRange(200, 400);
    /** When true, small images are enlarged to the defined size range. */
    private boolean minSizeRequired = false;

    /** Probability for returning no image from the GetNext() method. */
    private int yieldNullPercentage = 0;

    /** Synchronized queue for placing the loaded images. */
    private Queue<ContourImage> queue;

    /** Desired number of images that should be in the queue. */
    private final int queueLength;

    /** Background thread responsible for filling the queue. */
    private Thread thread;

    /** Used for regulating the buffered queue length. */
    private Semaphore queueEmptySemaphore;

    /** Used for regulating the buffered queue length. */
    private Semaphore queueFullSemaphore;

    /**
     * A list of recently used images.
     * We'll try to avoid using the same images in rapid succession.
     */
    private final List<String> recentlyUsedImages = new ArrayList<String>();

    //--------------------- Constructors

    /**
     * Constructor.
     * @param minSize Minimum size (in each dimension) the images should be scaled to.
     * @param maxSize Maximum size (in each dimension) the images should be scaled to.
     * @param minSizeRequired When true, small images are enlarged to the defined size range.
     * @param imageFolder This directory and its sub-directories are searched for images to be loaded.
     * @param queueLength Number of images to be pre-loaded into a queue.  If positive, the loader uses a background thread.
     * @param threadName Unique name of this image loader thread.  Only used for positive queueLength.
     */
    public ImageLoader(int minSize, int maxSize, boolean minSizeRequired, String imageFolder, int queueLength, String threadName)
    {
        this.sizeRange.min = minSize;
        this.sizeRange.max = maxSize;
        this.sizeRange.includingMax = true;
        this.minSizeRequired = minSizeRequired;
        this.imageFolder = imageFolder;
        this.queueLength = queueLength;

        if (queueLength > 0)
        {
            this.queue = new LinkedBlockingQueue<ContourImage>(queueLength);
            //this.queueEmptySemaphore = new Semaphore(0, queueLength);
            //this.queueFullSemaphore = new Semaphore(queueLength, queueLength);

            this.thread = new Thread(new Runnable()
            {
                @Override
                public void run()
                {
                    LoadImages();
                }
            });

            // The background thread should run with high priority until a list of image filenames
            // has been loaded (see FindImages()).
            thread.setPriority(Thread.MAX_PRIORITY);
            thread.setDaemon(true);
            thread.setName(threadName);
            thread.start();
        }
    }

    /**
     * @return An {@link ImageLoader} configured with the persistent screen saver parameters
     * or null if the "image number" option is 0.
     */
    public static ImageLoader GetScreenSaverImageLoader()
    {
        int count = RegisteredOptions.GetIntSetting(RegisteredOptions.OPT_IMAGE_NUMBER);
        int minSize = RegisteredOptions.GetIntSetting(RegisteredOptions.OPT_IMAGE_MIN_SIZE);
        int maxSize = RegisteredOptions.GetIntSetting(RegisteredOptions.OPT_IMAGE_MAX_SIZE);
        String imageFolder = RegisteredOptions.GetStringSetting(RegisteredOptions.OPT_IMAGE_FOLDER);
        boolean backgroundImages = RegisteredOptions.GetBoolSetting(RegisteredOptions.OPT_BACKGROUND_IMAGES);

        if (count > 0)
        {
            ImageLoader result = new ImageLoader(minSize, maxSize, false, imageFolder, count + 2, "FGIL");
            result.setYieldNullPercentage(backgroundImages ? 5 : 0);
            return result;
        }
        else
        {
            return null;
        }
    }

    //--------------------- IImageLoader implementation

    @Override
    public ContourImage getNext(Random r)
    {
        ContourImage result;

        if (yieldNullPercentage > 0 && r.nextInt(100) < yieldNullPercentage)
        {
            return null;
        }

        if (queueLength > 0)
        {
            result = Dequeue();
        }
        else
        {
            List<String> imagePaths = FindImages(imageFolder, 1, false, r);
            if (imagePaths.size() == 0)
            {
                return null;
            }
            String imagePath = imagePaths.get(0);
            result = LoadImage(imagePath, r);
            result.processImage();
        }

        return result;
    }

    @Override
    public void setYieldNullPercentage(int value)
    {
        this.yieldNullPercentage = value;
    }

    @Override
    public void shutdown()
    {
        if (this.thread != null)
        {
            try
            {
                thread.stop(); // TODO: set an indicator variable and interrupt the thread
                SaveImagePaths();
                thread.join();
            }
            catch (InterruptedException e)
            {
                // do nothing
            }
        }
    }

    //--------------------- Image provider thread.

    /**
     *
     */
    protected void LoadImages()
    {
        Random r = RandomFactory.createRandom();

        //--------------------- Start with a few images, preferably without a contour.

        // The saved image paths are already ordered (see SaveImagePaths()).
        LoadAndEnqueue(r, true, LoadImagePaths());

        if (queue.size() >= queueLength)
        {
            ReduceThreadPriority();
        }
        else
        {
            LoadAndEnqueue(r, false, FindImages(imageFolder, queueLength + 1, true, r));
        }

        //--------------------- Continuously load more images, keeping the queue full.

        while (true)
        {
            int loadedImagesCount = 0;

            for (String imagePath : FindImages(imageFolder, 100, false, r))
            {
                ContourImage img = LoadImage(imagePath, r);
                if (img == null)
                {
                    continue;
                }

                Enqueue(img);
                ++loadedImagesCount;
            }

            // If no image was loaded successfully, enqueue a null value.
            if (loadedImagesCount == 0)
            {
                Enqueue(null);
            }
        }
    }

    /**
     * @param r A source of random numbers.
     * @param preserveOrder When false, the list is re-ordered so that images without a contour that need not be processed come first.
     * @param imagePaths A list of fully qualified image file paths.
     */
    private void LoadAndEnqueue(Random r, boolean preserveOrder, Iterable<String> imagePaths)
    {
        //Log.WriteLine("{ " + string.Format("LoadAndEnqueue({0})", imagePaths.GetType().ToString()));
        List<ContourImage> unprocessedImages = new ArrayList<ContourImage>(queueLength + 1);

        Iterator<String> e = imagePaths.iterator();
        while(e.hasNext())
        {
            String imagePath = e.next();
            ContourImage img = null;

            if (imagePath != null)
            {
                img = LoadImage(imagePath, r);
            }

            if (img == null)
            {
                // continue;
            }
            else if (!preserveOrder && img.hasContour())
            {
                //Log.WriteLine("- LoadAndEnqueue() save contour image: " + img.ToString());
                unprocessedImages.add(img);
            }
            else
            {
                Enqueue(img);
            }

            e.remove();
        }

        for (ContourImage img : unprocessedImages)
        {
            Enqueue(img);
        }
        //Log.WriteLine("} LoadAndEnqueue()");
    }


    /**
     * @param imagePath A fully qualified image file path.
     * @param r A source of random numbers.
     * @return An image loaded from the given path and scaled to the known size range.
     */
    private ContourImage LoadImage(String imagePath, Random r)
    {
        try
        {
            //Log.WriteLine("{ " + string.Format("LoadImage({0})", imagePath));
            BufferedImage img = ImageIO.read(new File(imagePath));

            //Log.WriteLine("- LoadImage() scale image");
            // [start] Scale img so that its larger dimension is between the desired min and max size.

            boolean resize = false;

            if (img.getWidth() > sizeRange.max || img.getHeight() > sizeRange.max)
            {
                resize = true;
            }
            if (minSizeRequired && img.getWidth() < sizeRange.min && img.getHeight() < sizeRange.min)
            {
                resize = true;
            }

            if (resize)
            {
                int d = sizeRange.pick(r);
                int h = img.getHeight(), w = img.getWidth();
                if (h > w)
                {
                    w = d * w / h;
                    h = d;
                }
                else
                {
                    h = d * h / w;
                    w = d;
                }

                BufferedImage tmpImg = ImageUtil.resize(img, w, h);
                img = tmpImg;
            }

            // [end]

            //Log.WriteLine("- LoadImage() create contour image");
            ContourImage result = new ContourImage(img, imagePath);

            //Log.WriteLine("} LoadImage()");
            return result;
        }
        catch (Exception e)
        {
            System.out.println(String.format("failed loading image [%s]: %s", imagePath, e.toString()));
            return null;
        }
    }


    /**
     * Processes and enqueues the given image.
     * @param img A {@link ContourImage} that has not yet been processed.
     */
    private void Enqueue(ContourImage img)
    {
        //Log.WriteLine("{ " + string.Format("Enqueue({0})", img.ToString()));

        // Wait while the queue contains enough items.
        queueFullSemaphore.acquireUninterruptibly();

        if (img != null)
        {
            img.processImage();
        }

        // Enqueue the processed image.
        //Log.WriteLine("- Enqueue() Enqueue");
        queue.add(img);
        queueEmptySemaphore.release();

        if (true)
        {
            /* Continue at the lower thread priority as soon as the first image has been loaded.
             * This assumes that a) we need only one image or b) we can easily load more images on demand.
             * This works fine if enough images can be loaded from the saved images list.
             * See LoadImagePaths().
             */
            ReduceThreadPriority();
        }

        //Log.WriteLine("} Enqueue()");
    }


    /**
     * @return The next {@link ContourImage} from the queue.
     */
    private ContourImage Dequeue()
    {
        //Log.WriteLine("{ Dequeue()");
        queueEmptySemaphore.acquireUninterruptibly();
        //Log.WriteLine("- Dequeue() Dequeue");
        ContourImage result = queue.poll();
        queueFullSemaphore.release();
        //Log.WriteLine("} Dequeue()");
        return result;
    }


    /**
     * @param folderPath The full path to a directory containing image files.
     * @param count Number of file names that should be collected.
     * @param quickSearch When true, only JPG files are considered (if there are at least 100 of them).
     * @param r A source of random numbers.
     * @return A list of file names in or below the given path.
     * Only the following image file types are considered: JPG, GIF, PNG.
     */
    private List<String> FindImages(String folderPath, int count, boolean quickSearch, Random r)
    {
        //Log.WriteLine("{ FindImages()");
        if (folderPath == null || count < 1)
        {
            return new ArrayList<String>();
        }

        List<String> availableImages = new ArrayList<String>();

        availableImages.addAll(swa.util.Directory.Find(folderPath, "*.jpg", true));

        if (quickSearch == false || availableImages.size() < 100)
        {
            availableImages.addAll(swa.util.Directory.Find(folderPath, "*.gif", true));
            availableImages.addAll(swa.util.Directory.Find(folderPath, "*.png", true));
        }

        // Initially, the thread was started with high priority (see the Constructor).
        // After the list has been loaded, it shall continue with low priority.
        ReduceThreadPriority();

        ArrayList<String> result = new ArrayList<String>(count);

        // Shorten the list of recently used images.
        // Make sure the list does not get too short.
        while (this.recentlyUsedImages.size() > 0
            && this.recentlyUsedImages.size() > availableImages.size() - count
            && this.recentlyUsedImages.size() > availableImages.size() * 3 / 4
            )
        {
            // Remove an item near the beginning of the list (recently added items are at the end).
            this.recentlyUsedImages.remove(r.nextInt(this.recentlyUsedImages.size() / 3 + 1));
        }

        // Select the required number of images.
        // Avoid recently used images.
        while (result.size() < count && availableImages.size() > 0)
        {
            int p = r.nextInt(availableImages.size());
            String imagePath = availableImages.get(p);

            if (!recentlyUsedImages.contains(imagePath))
            {
                result.add(imagePath);
                recentlyUsedImages.add(imagePath);
            }

            availableImages.remove(p);
        }

        //Log.WriteLine("} FindImages()");
        return result;
    }


    /**
     * Sets the background thread priority to a minimum.
     */
    private void ReduceThreadPriority()
    {
        int prio = Thread.MIN_PRIORITY;

        if (this.thread != null && thread.getPriority() != prio)
        {
            //Log.WriteLine("{ ReduceThreadPriority() set priority");
            thread.setPriority(prio);
            //Thread.sleep(0);
            //Log.WriteLine("} ReduceThreadPriority()");
        }
    }

    //---------------------  Persistent memory of image paths.


    //--------------------- Persistent memory of image paths.

    /**
     * Writes the paths of all currently queued images to the registry.
     * These paths will be used to initialize the queue on the following run.
     */
    private void SaveImagePaths()
    {
        //Log.WriteLine("{ SaveImagePaths()");

        // [start] Get the paths of all currently queued images.

        List<String> paths = new ArrayList<String>(queue.size());

        while (queue.size() > 0)
        {
            ContourImage img = Dequeue();

            if (img.getPath() != null)
            {
                if (img.hasContour())
                {
                    // True contour images should be loaded last.
                    paths.add(img.getPath());
                }
                else
                {
                    // Images without a contour should be loaded first.
                    paths.add(0, img.getPath());
                }
            }
        }

        // [end]

        // [start] Concatenate all paths.

        StringBuilder s = new StringBuilder(2048);

        for (String path : paths)
        {
            //Log.WriteLine("- SaveImagePaths() " + path);
            if (s.length() > 0) { s.append(PathSeparator); }
            s.append(path.substring(imageFolder.length()));
        }

        // [end]

        IParameterFile params = RegisteredOptions.getParameterFile();
        if (params != null)
        {
            params.set(thread.getName() + " " + RegisteredOptions.SAVE_IMAGE_PATHS, s.toString());
        }
        //Log.WriteLine("} SaveImagePaths()");
    }


    /**
     * @return A list of image path names that have been stored in the previous run.
     */
    private Iterable<String> LoadImagePaths()
    {
        //Log.WriteLine("{ LoadImagePaths()");
        String s = RegisteredOptions.GetStringSetting(thread.getName() + " " + RegisteredOptions.SAVE_IMAGE_PATHS);
        String[] paths = s.split(PathSeparator);

        for (int i = 0; i < paths.length; i++)
        {
            paths[i] = imageFolder + paths[i];
        }

        List<String> result = Arrays.asList(paths);
        recentlyUsedImages.addAll(result);

        //Log.WriteLine("} LoadImagePaths()");
        return result;
    }
}
